url29ff2e53f3_nicolas3470_Settlers_of_Catan_AI_tgz-pJ8-robotGameJ8/src/soc/robot/SOCRobotDM.java

```diff
@@ -0,0 +1,2114 @@
+	queue.clear();
+	Iterator necRoadsIter = necRoadVec.iterator();
+	while (necRoadsIter.hasNext()) {
+	  SOCPossibleRoad necRoad = (SOCPossibleRoad)necRoadsIter.next();
+	  D.ebugPrintln("-- queuing necessary road at "+game.getBoard().edgeCoordToString(necRoad.getCoordinates()));
+	  queue.put(new Pair(necRoad, null));
+	}
+	//
+	// Do a BFS of the necessary road paths looking for the shortest one.
+	//
+	while (!queue.empty()) {
+	  Pair dataPair = (Pair)queue.get();
+	  SOCPossibleRoad curRoad = (SOCPossibleRoad)dataPair.getA();
+	  D.ebugPrintln("-- current road at "+game.getBoard().edgeCoordToString(curRoad.getCoordinates()));
+	  Vector necRoads = curRoad.getNecessaryRoads();
+	  if (necRoads.isEmpty()) {
+	    //
+	    // we have a path 
+	    //
+	    D.ebugPrintln("Found a path!");
+	    Stack path = new Stack();
+	    path.push(curRoad);
+	    Pair curPair = (Pair)dataPair.getB();
+	    D.ebugPrintln("curPair = "+curPair);
+	    while (curPair != null) {
+	      path.push(curPair.getA());
+	      curPair = (Pair)curPair.getB();
+	    }
+	    posSet.setRoadPath(path);
+	    queue.clear();
+	    D.ebugPrintln("Done setting path.");
+	  } else {
+	    necRoadsIter = necRoads.iterator();
+	    while (necRoadsIter.hasNext()) {
+	      SOCPossibleRoad necRoad2 = (SOCPossibleRoad)necRoadsIter.next();
+	      D.ebugPrintln("-- queuing necessary road at "+game.getBoard().edgeCoordToString(necRoad2.getCoordinates()));
+	      queue.put(new Pair(necRoad2, dataPair));
+	    }
+	  }
+	}
+	D.ebugPrintln("Done searching for path.");
+
+	//
+	// calculate ETA
+	//
+	SOCResourceSet targetResources = new SOCResourceSet();
+	targetResources.add(SOCGame.SETTLEMENT_SET);
+	int pathLength = 0;
+	Stack path = posSet.getRoadPath();
+	if (path != null) {
+	  pathLength = path.size();
+	}
+	for (int i = 0; i < pathLength; i++) {
+	  targetResources.add(SOCGame.ROAD_SET);
+	}
+	try {
+	  SOCResSetBuildTimePair timePair = ourBSE.calculateRollsFast(ourPlayerData.getResources(), targetResources, 100, ourPlayerData.getPortFlags());
+	  posSet.setETA(timePair.getRolls());
+	} catch (CutoffExceededException ex) {
+	  posSet.setETA(100);
+	}
+      } else {
+	//
+	// no roads are necessary
+	//
+	posSet.setRoadPath(null);
+	posSet.setETA(settlementETA);
+      }
+      D.ebugPrintln("Settlement ETA = "+posSet.getETA());
+    }
+  }
+
+  /**
+   * Does a depth first search from the end point of the longest
+   * path in a graph of nodes and returns how many roads would 
+   * need to be built to take longest road.
+   *
+   * @param startNode     the path endpoint
+   * @param pathLength    the length of that path
+   * @param lrLength      length of longest road in the game
+   * @param searchDepth   how many roads out to search
+   *
+   * @return a stack containing the path of roads with the last one on top, or null if it can't be done
+   */
+  private Stack recalcLongestRoadETAAux(int startNode, int pathLength, int lrLength, int searchDepth) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 46. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 1099-1149 into a separate method.